client/pd_service_discovery.go

```diff
@@ -837,42 +950,106 @@ func (c *pdServiceDiscovery) updateURLs(members []*pdpb.Member) {
+func (c *pdServiceDiscovery) switchLeader(addrs []string) (bool, error) {
 	// FIXME: How to safely compare leader urls? For now, only allows one client url.
 	addr := addrs[0]
-	oldLeader := c.getLeaderAddr()
-	if addr == oldLeader {
-		return nil
+	oldLeader := c.getLeaderServiceClient()
+	if addr == oldLeader.GetAddress() && oldLeader.GetClientConn() != nil {
+		return false, nil
 	}
 
-	if _, err := c.GetOrCreateGRPCConn(addr); err != nil {
-		log.Warn("[pd] failed to connect leader", zap.String("leader", addr), errs.ZapError(err))
+	newConn, err := c.GetOrCreateGRPCConn(addr)
+	// If gRPC connect is created successfully or leader is new, still saves.
+	if addr != oldLeader.GetAddress() || newConn != nil {
+		// Set PD leader and Global TSO Allocator (which is also the PD leader)
+		leaderClient := newPDServiceClient(addr, addr, newConn, true)
+		c.leader.Store(leaderClient)
 	}
 	// Set PD leader and Global TSO Allocator (which is also the PD leader)
 	c.leader.Store(addr)
 	// Run callbacks
 	if c.tsoGlobalAllocLeaderUpdatedCb != nil {
 		if err := c.tsoGlobalAllocLeaderUpdatedCb(addr); err != nil {
-			return err
+			return true, err
 		}
 	}
 	for _, cb := range c.leaderSwitchedCbs {
 		cb()
 	}
-	log.Info("[pd] switch leader", zap.String("new-leader", addr), zap.String("old-leader", oldLeader))
-	return nil
+	log.Info("[pd] switch leader", zap.String("new-leader", addr), zap.String("old-leader", oldLeader.GetAddress()))
+	return true, err
 }
 
-func (c *pdServiceDiscovery) updateFollowers(members []*pdpb.Member, leader *pdpb.Member) {
-	var addrs []string
+func (c *pdServiceDiscovery) updateFollowers(members []*pdpb.Member, leader *pdpb.Member) (changed bool) {
+	followers := make(map[string]*pdServiceClient)
+	c.followers.Range(func(key, value any) bool {
+		followers[key.(string)] = value.(*pdServiceClient)
+		return true
+	})
+	var followerAddrs []string
 	for _, member := range members {
 		if member.GetMemberId() != leader.GetMemberId() {
 			if len(member.GetClientUrls()) > 0 {
-				addrs = append(addrs, member.GetClientUrls()...)
+				followerAddrs = append(followerAddrs, member.GetClientUrls()...)
+
+				// FIXME: How to safely compare urls(also for leader)? For now, only allows one client url.
+				addr := member.GetClientUrls()[0]
+				if client, ok := c.followers.Load(addr); ok {
+					if client.(*pdServiceClient).GetClientConn() == nil {
+						conn, err := c.GetOrCreateGRPCConn(addr)
+						if err != nil || conn == nil {
+							log.Warn("[pd] failed to connect follower", zap.String("follower", addr), errs.ZapError(err))
+							continue
+						}
+						follower := newPDServiceClient(addr, leader.GetClientUrls()[0], conn, false)
+						c.followers.Store(addr, follower)
+						changed = true
+					}
+					delete(followers, addr)
+				} else {
+					changed = true
+					conn, err := c.GetOrCreateGRPCConn(addr)
+					follower := newPDServiceClient(addr, leader.GetClientUrls()[0], conn, false)
```

Comment:
leader?

Reply:
Yes, it will save the address of  leader for forward.

Reply:
Is it possible len(GetClientUrls)==0?

Reply:
No, it will return will before here